### ISSUE TITLE
Build Profiling and Dynamic Ways for libraries

### ISSUE THREAD

**Issue by chitrak7** [snowleopard/hadrian#626]
We need Profiling and Dynamic Ways to pass validation. These ways are required for ghci, base and rts.
@snowleopard  @alpmestan

**Comment by chitrak7** [comment]
I tried doing this but this is not working.
`needTestsuiteBuilders :: Action ()
needTestsuiteBuilders = do
    targets <- mapM (needfile Stage1) =<< testsuitePackages
    need targets
  where
    needfile :: Stage -> Package -> Action FilePath
    needfile stage pkg
      | isLibrary pkg = pkgConfFile (Context stage pkg profilingDynamic)
      | otherwise = programPath =<< programContext stage pkg
`

**Comment by chitrak7** [comment]
`SUMMARY for test run started at Sat Jun 16 00:37:07 2018 IST`
` 0:19:36 spent to go through`
 `   6418 total tests, which gave rise to`
 `  23606 test cases, of which`
 `  17251 were skipped`
` `
 `     28 had missing libraries`
 `   6178 expected passes`
 `     91 expected failures`
` `
 `      0 caused framework failures`
 `      0 caused framework warnings`
 `      0 unexpected passes`
 `     56 unexpected failures`
  `     2 unexpected stat failures`
The missing libraries are due to dynamic and profiling and maybe threaded libraries. Resolving this issue will hopefully bring the failures below 20.

**Comment by chitrak7** [comment]
`/home/chitrak/ghc/_build/stage1/lib/../lib/x86_64-linux-ghc-8.5.20180616/rts-1.0/libHSrts-1.0_thr_l.a(GC.thr_l_o)(.text+0xb): error: undefined reference to 'evacuate'
/home/chitrak/ghc/_build/stage1/lib/../lib/x86_64-linux-ghc-8.5.20180616/rts-1.0/libHSrts-1.0_thr_l.a(GC.thr_l_o)(.text+0x43): error: undefined reference to 'scavenge_loop'
/home/chitrak/ghc/_build/stage1/lib/../lib/x86_64-linux-ghc-8.5.20180616/rts-1.0/libHSrts-1.0_thr_l.a(GC.thr_l_o):function GarbageCollect: error: undefined reference to 'scavenge_capability_mut_lists'
/home/chitrak/ghc/_build/stage1/lib/../lib/x86_64-linux-ghc-8.5.20180616/rts-1.0/libHSrts-1.0_thr_l.a(GC.thr_l_o):function GarbageCollect: error: undefined reference to 'scavenge_capability_mut_lists'
/home/chitrak/ghc/_build/stage1/lib/../lib/x86_64-linux-ghc-8.5.20180616/rts-1.0/libHSrts-1.0_thr_l.a(GC.thr_l_o):function gcWorkerThread: error: undefined reference to 'scavenge_capability_mut_lists'
/home/chitrak/ghc/_build/stage1/lib/../lib/x86_64-linux-ghc-8.5.20180616/rts-1.0/libHSrts-1.0_thr_l.a(MarkWeak.thr_l_o):function markWeakPtrList: error: undefined reference to 'evacuate'
/home/chitrak/ghc/_build/stage1/lib/../lib/x86_64-linux-ghc-8.5.20180616/rts-1.0/libHSrts-1.0_thr_l.a(MarkWeak.thr_l_o):function scavengeLiveWeak: error: undefined reference to 'evacuate'
/home/chitrak/ghc/_build/stage1/lib/../lib/x86_64-linux-ghc-8.5.20180616/rts-1.0/libHSrts-1.0_thr_l.a(MarkWeak.thr_l_o):function scavengeLiveWeak: error: undefined reference to 'evacuate'
collect2: error: ld returned 1 exit status
`gcc' failed in phase `Linker'. (Exit code: 1)

Currently, atleast 60% of erros show tis error message. I am stuck at where to go with it.
`

**Comment by snowleopard** [comment]
@chitrak7 @alpmestan Perhaps, there should be a special `test` build flavour? This may simplify the logic for setting up the right ways, as well as selecting the right collection of packages to build.

**Comment by chitrak7** [comment]
But still, by changing the context of libraries, these ways should be built. But it is not. 

> needfile stage pkg | isLibrary pkg = pkgConfFile (Context stage pkg profilingDynamic)

This is how I modified the rule for libraries needed. But still dynamic ways were not supported.

**Comment by snowleopard** [comment]
I'm copying my comment in one of the PRs here:

> I think someone needs to think about the overall architecture of the testsuite. So far there are a bunch of PRs that seem to introduce various ad hoc solutions, but I do not see a clear design of how everything should be tied together. Can you write up such a design document?

**Comment by alpmestan** [comment]
Is this really addressed yet? I don't think so.

**Comment by snowleopard** [comment]
Oops, I think this got auto-closed when merging the PR.

**Comment by chitrak7** [comment]
@alpmestan @angerman I still couldn't figure out how Hadrian makes flags while running cabal. I believe that the flags are built based on the flavor supplied and not depending on the package context, causing errors.
@snowleopard Can you look into the dead code that you mentioned. Removing it can make the process much simpler.

**Comment by alpmestan** [comment]
> @alpmestan @angerman I still couldn't figure out how Hadrian makes flags while running cabal. I believe that the flags are built based on the flavor supplied and not depending on the package context, causing.

Have you taken a look at `Settings.Builders.GhcCabal` ?

**Comment by chitrak7** [comment]
@alpmestan Yes, I did. I made some changes in this file in #628 also. But they didn't seem to work.

**Comment by alpmestan** [comment]
I need some more details to be able to help you out here. Could you perhaps isolate those changes and show us the patch, telling us what did not work as expected?

**Comment by snowleopard** [comment]
> Have you taken a look at `Settings.Builders.GhcCabal`?

@alpmestan Isn't this just dead code? We got rid of the dependency on `ghc-cabal`, so we no longer need this particular file, hence I was planning to remove it, but haven't got around to it yet.

**Comment by angerman** [comment]
I doubt that `GhcCabal` in itself is dead. There is some unfortunate(?) reuse of the `GhcCabal` builder for argument collection.  The actual configuration happens in
https://github.com/snowleopard/hadrian/blob/f3192439950ba3a3af58410c331d2492322dd4ab/src/Hadrian/Haskell/Cabal/Parse.hs#L118-L160

In general the `ghcCabal` package is dead, as well as any `ghc-cabal`/`package-data.mk` logic.

**Comment by snowleopard** [comment]
@angerman I think we should drop the `GhcCabal` builder and move all Cabal-related arguments to `CabalFlags`. That would make it easier to understand what's going on.

**Comment by angerman** [comment]
Well, we could rename `GhcCabal` to `CabalArgs` or something, but `CabalFlags` represents what `cabal` calls `flags`, e.g. `-fXXX`, or `-f-YYY`.

So we won't be able to flatten both.

**Comment by snowleopard** [comment]
Well, we won't be able to flatten expressions, but I don't think we need a separate builder. Anyway, I will look into this -- this has been on my todo list for a while.

**Comment by chitrak7** [comment]
@alpmestan This is how I tried to change the code for GhcCabal. Earlier, it would only look at flavour ways, but I added contextWay option as well. Looks like unfortunately this is not working.
```libraryArgs :: Args
libraryArgs = do
    flavourWays <- getLibraryWays
    contextWay  <- getWay
    withGhci    <- expr ghcWithInterpreter
    dynPrograms <- dynamicGhcPrograms <$> expr flavour
    let ways = flavourWays ++ [contextWay]
    pure [ if vanilla `elem` ways
           then  "--enable-library-vanilla"
           else "--disable-library-vanilla"
         , if vanilla `elem` ways && withGhci && not dynPrograms
           then  "--enable-library-for-ghci"
           else "--disable-library-for-ghci"
         , if or [Profiling `wayUnit` way | way <- ways]
           then  "--enable-library-profiling"
           else "--disable-library-profiling"
         , if or [Dynamic `wayUnit` way | way <- ways]
           then  "--enable-shared"
           else "--disable-shared" ]
```

@snowleopard We are only using GhcCabal once, that too only to compute args. It is better to remove the builder and make args separately. Will certainly reduce the confusion. Should I create an issue for this?

**Comment by snowleopard** [comment]
@chitrak7 Yes, I think the removal of `GhcCabal` builder deserves a separate issue. Please create one!

**Comment by snowleopard** [comment]
The issue with `GhcCabal` and `CabalFlags` builders has been addressed in #658.

**Comment by snowleopard** [comment]
I believe this has been fixed.
